pkg/plugin/ciliumeventobserver/ciliumeventobserver_linux.go

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+// Package ciliumeventobserver contains the Retina CiliumEventObserver plugin. It uses unix socket to get events from cilium and decode them to flow objects.
+package ciliumeventobserver
+
+import (
+	"context"
+	"encoding/gob"
+	"errors"
+	"io"
+	"net"
+	"time"
+
+	v1 "github.com/cilium/cilium/pkg/hubble/api/v1"
+	"github.com/cilium/cilium/pkg/monitor/payload"
+	kcfg "github.com/microsoft/retina/pkg/config"
+	"github.com/microsoft/retina/pkg/enricher"
+	"github.com/microsoft/retina/pkg/log"
+	"github.com/microsoft/retina/pkg/plugin/api"
+	"go.uber.org/zap"
+)
+
+const (
+	defaultAttempts   = 5
+	defaultRetryDelay = 12 * time.Second
+	workers           = 2
+)
+
+var (
+	errFailedConnection = errors.New("Failed to connect to cilium monitor after max attempts")
+	errFailedToDial     = errors.New("Failed to dial cilium monitor socket")
+)
+
+func New(cfg *kcfg.Config) api.Plugin {
+	return &ciliumeventobserver{
+		cfg:           cfg,
+		l:             log.Logger().Named(string(Name)),
+		retryDelay:    defaultRetryDelay,
+		maxAttempts:   defaultAttempts,
+		sockPath:      cfg.MonitorSockPath,
+		payloadEvents: make(chan *payload.Payload),
+		d:             &DefaultDialer{},
+	}
+}
+
+func (c *ciliumeventobserver) Name() string {
+	return string(Name)
+}
+
+func (c *ciliumeventobserver) Generate(_ context.Context) error {
+	return nil
+}
+
+func (c *ciliumeventobserver) Compile(_ context.Context) error {
+	return nil
+}
+
+func (c *ciliumeventobserver) Init() error {
+	c.p = &parser{
+		l: c.l,
+	}
+	err := c.p.Init()
+	if err != nil {
+		c.l.Error("Failed to initialize ciliumeventobserver parser", zap.Error(err))
+		return err
+	}
+	c.l.Info("Initialized ciliumeventobserver plugin")
+	return nil
+}
+
+func (c *ciliumeventobserver) Start(ctx context.Context) error {
+	if c.cfg.EnablePodLevel {
+		if enricher.IsInitialized() {
+			c.enricher = enricher.Instance()
+		} else {
+			c.l.Warn("retina enricher is not initialized")
+		}
+	}
+
+	for i := 0; i < workers; i++ {
+		c.wg.Add(1)
```

Comment:
Do you need the waitgroup? The threads will exit once main context is cancelled.